c10/cuda/CUDAMallocAsyncAllocator.cpp

```diff
@@ -0,0 +1,823 @@
+#include <c10/cuda/CUDACachingAllocator.h>
+#include <c10/cuda/CUDAException.h>
+#include <c10/cuda/CUDAFunctions.h>
+#include <c10/cuda/CUDAGuard.h>
+#include <c10/util/UniqueVoidPtr.h>
+#include <c10/util/irange.h>
+
+#include <mutex>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace c10 {
+namespace cuda {
+namespace CUDACachingAllocator {
+namespace CudaMallocAsync {
+
+#if CUDA_VERSION >= 11040
+// CUDA device allocator that uses cudaMallocAsync to implement
+// the same interface as CUDACachingAllocator.cpp.
+
+// Designed to be safe for CUDA graph capture.
+// Interactions with CUDA graph capture are mediated by
+// notifyCaptureBegin
+// notifyCaptureAboutToEnd
+// notifyCaptureEnded
+// notifyCaptureDestroy
+
+// Implementation details, not declared in CUDACachingAllocator.h
+namespace {
+
+// General helpers
+
+struct UsageStream {
+  cudaStream_t stream;
+  int device;
+  UsageStream() {}
+  UsageStream(cudaStream_t s, int d) : stream(s), device(d) {}
+  UsageStream(const UsageStream& us) : stream(us.stream), device(us.device) {}
+  UsageStream(const UsageStream&& us) : stream(us.stream), device(us.device) {}
+  UsageStream& operator=(UsageStream other) {
+    stream = other.stream;
+    device = other.device;
+    return *this;
+  }
+};
+
+struct PtrUsage {
+  std::vector<UsageStream> usage_streams;
```

Comment:
It seems to me that this should be a set, not a vector. Although the number of usage streams will be typically small, a user may call recordStream multiple times for an allocation which means this vector could grow unboundedly. The original implementation didn't have this problem because we recorded stream uses as a set.

Reply:
It's also important for the pointer to remember which of its usage streams was the original, though. With a vector that's convenient because it's just usage_streams[0]. I guess I could change this to ska::flat_hash_set and stash the original usage stream in a separate, dedicated field.

Reply:
Yeah do that

Reply:
done, also i avoid using the hash set at all unless there are recorded streams beyond the initial creation stream.